File: app/src/main/java/com/example/david/lists/di/data/AppComponent.java
package com.example.david.lists.di.data;

import android.app.Application;
import android.content.SharedPreferences;

import com.example.david.lists.data.repository.IRepository;
import com.example.david.lists.data.repository.IUserRepository;

import javax.inject.Singleton;

import dagger.BindsInstance;
import dagger.Component;

@Singleton
@Component(modules = {
        RepositoryModule.class,
        UserRepositoryModule.class,
        RemoteRepositoryModule.class,
        SharedPreferencesModule.class
})
public interface AppComponent {
    IRepository repository();

    IUserRepository userRepository();

    SharedPreferences sharedPrefsNightMode();

    @Component.Builder
    interface Builder {
        AppComponent build();

        @BindsInstance
        Builder application(Application application);
    }
}

File: app/src/main/java/com/precopia/david/lists/common/buildlogic/AppComponent.kt
package com.precopia.david.lists.common.buildlogic

import android.app.Application
import android.content.SharedPreferences
import com.precopia.david.lists.util.IUtilThemeContract
import com.precopia.domain.repository.IRepositoryContract
import dagger.BindsInstance
import dagger.Component
import javax.inject.Singleton

@Singleton
@Component(modules = [
    RepositoryModule::class,
    SharedPrefsModule::class,
    UtilThemeModule::class
])
interface AppComponent {
    fun utilTheme(): IUtilThemeContract

    fun sharedPrefs(): SharedPreferences

    fun repo(): IRepositoryContract.Repository

    fun userRepo(): IRepositoryContract.UserRepository

    @Component.Builder
    interface Builder {
        fun build(): AppComponent

        @BindsInstance
        fun application(application: Application): Builder
    }
}
